### DISCUSSION TITLE
Pyo3 builds on Mac M1 builds failing with undefined symbols for arm64 libintl

### DISCUSSION THREAD

**Discussion by twitu** [PyO3/pyo3#2823] [Questions] [answered]
I can build a pyo3 project on Mac M1 but when I run or test it, it fails to load or find symbols.

```
          ld: warning: object file (/Users/twitu/Code/nautilus_trader/nautilus_core/target/debug/deps/libpyo3_ffi-77a14a70a5d0a67c.rlib(accu.o)) was built for newer macOS version (12.3) than being linked (12.0)
          Undefined symbols for architecture arm64:
            "_libintl_bind_textdomain_codeset", referenced from:
                __locale_bind_textdomain_codeset in libpyo3_ffi-77a14a70a5d0a67c.rlib(_localemodule.o)
            "_libintl_bindtextdomain", referenced from:
                __locale_bindtextdomain in libpyo3_ffi-77a14a70a5d0a67c.rlib(_localemodule.o)
            "_libintl_textdomain", referenced from:
                __locale_textdomain in libpyo3_ffi-77a14a70a5d0a67c.rlib(_localemodule.o)
            "_libintl_dcgettext", referenced from:
                __locale_dcgettext in libpyo3_ffi-77a14a70a5d0a67c.rlib(_localemodule.o)
            "_libintl_setlocale", referenced from:
                __locale_setlocale in libpyo3_ffi-77a14a70a5d0a67c.rlib(_localemodule.o)
                __locale_localeconv in libpyo3_ffi-77a14a70a5d0a67c.rlib(_localemodule.o)
            "_libintl_dgettext", referenced from:
                __locale_dgettext in libpyo3_ffi-77a14a70a5d0a67c.rlib(_localemodule.o)
            "_libintl_gettext", referenced from:
                __locale_gettext in libpyo3_ffi-77a14a70a5d0a67c.rlib(_localemodule.o)
          ld: symbol(s) not found for architecture arm64
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

In Mac M1, homebrew has a slightly different prefix. So I had to pass `export RUSTFLAGS='-L /opt/homebrew/lib'`. Before passing the flag the error is even less descriptive.

```
dyld[5845]: symbol not found in flat namespace '_libintl_bind_textdomain_codeset'
```

pyo3 version 0.17.3
rustc version 1.66.0

**Comment by messense** [comment, marked as answer]
> ld: warning: object file (/Users/twitu/Code/nautilus_trader/nautilus_core/target/debug/deps/libpyo3_ffi-77a14a70a5d0a67c.rlib(accu.o)) was built for newer macOS version (12.3) than being linked (12.0)

Have you tried `cargo clean` first?

`_libintl_*` symbols are from `gettext`, right? What's the output of `file /opt/homebrew/lib/libintl.8.dylib`? Mine is

```bash
$ file /opt/homebrew/lib/libintl.8.dylib
/opt/homebrew/lib/libintl.8.dylib: Mach-O 64-bit dynamically linked shared library arm64
```

**Reply by twitu** [reply]
I've tried cargo clean it gives the same result.

Here's the output.
```
 file /opt/homebrew/lib/libintl.8.dylib
/opt/homebrew/lib/libintl.8.dylib: Mach-O 64-bit dynamically linked shared library arm64
```

**Reply by messense** [reply]
Do you know where is the `gettext` dependency comes from? What CPython C-API used that requires it in your code? Did it feed `gettext` libraries to linker?

**Reply by twitu** [reply]
I'm not sure how to find this. I followed a couple of answers from this [SO question](https://stackoverflow.com/questions/73610525/how-to-find-out-which-rust-dependency-added-a-dynamically-linked-library).

```
cargo clean && cargo build -vv 2>/dev/null | grep 'rustc-link-lib'
```

This returns nothing for my binary that's failing to run.

```
readelf -d target/debug/<the-test-path> | grep NEEDED
```

And this finds that the built debug file for test is not of ELF format.

Doesn't the log already tell that it's trying to resolve the symbol in libpyo3 and is unable to find it. What am I specifically trying to find?

```
            "_libintl_setlocale", referenced from:
                __locale_setlocale in libpyo3_ffi-77a14a70a5d0a67c.rlib(_localemodule.o)
                __locale_localeconv in libpyo3_ffi-77a14a70a5d0a67c.rlib(_localemodule.o)
```

**Reply by messense** [reply]
> but when I run or test it, it fails to load or find symbols

It's not clear to me what you were trying to run/test, an extension module or a Rust `#[test]`? If it's the later, did you enable the `auto-initialize` feature and disable the `extension-module feature? See https://pyo3.rs/v0.17.3/faq.html#i-cant-run-cargo-test-or-i-cant-build-in-a-cargo-workspace-im-having-linker-issues-like-symbol-not-found-or-undefined-reference-to-_pyexc_systemerror

Without a repro, we are just speculating the cause of the issue, so it will be very helpful if you can provide more context, namely:

* How did you install python? Did you use a virtualenv?
* How did you run/test it?

**Reply by twitu** [reply]
Pardon the delay I was away for a bit.

> How did you install python? Did you use a virtualenv?

Python (3.10) is installed using pyenv (2.3.4) and managed using poetry (1.3.1).

I also tried to check if the [error is related to linking in virtual environments](https://github.com/PyO3/pyo3/issues/1741#issuecomment-888055873) but there is not binary called print-config anymore.

> How did you run/test it?

I tested it by running `cargo test` command in a shell where the python virtualenv (poetry shell) is activated.

Is there any other information I can share?

**Reply by messense** [reply]
Make sure your Python installed using pyenv is configured with `--enable-shared`:

```bash
PYTHON_CONFIGURE_OPTS="--enable-shared" pyenv install 3.10.9
```

See https://pyo3.rs/v0.17.3/getting_started#virtualenvs

**Reply by twitu** [reply]
I also tried to make a smaller test to reproduce and surprisingly even an empty test (with an empty body) fails with the same linking error.

**Reply by messense** [reply]
Can you paste the output of `python3 -m sysconfig`?

**Reply by twitu** [reply]
> Make sure your Python installed using pyenv is configured with `--enable-shared`:
> 
> ```shell
> PYTHON_CONFIGURE_OPTS="--enable-shared" pyenv install 3.10.9
> ```
> 
> See https://pyo3.rs/v0.17.3/getting_started#virtualenvs

Installing with this sharing enabled worked. Now the tests are passing and there is no linking error.

Thanks a lot for pointing out the fix. I should have read the documentation properly.

**Reply by twitu** [reply]
Also this flag `export RUSTFLAGS='-L /opt/homebrew/lib'` is NOT required after installing python with sharing enabled. The tests still work without it.
